### ISSUE TITLE
variants Sending in malformed reference name causes 500

### ISSUE THREAD

**Issue by david4096** [ga4gh/server#942]
If I send in a number as opposed to string to `variants/search` for the reference name I get a 500.

```
    payload = json.dumps({
        "variantSetId": variant_set['id'],
        "start": 0,           # genomic position
        "end": 2**32,  # genomic position
        "referenceName": 1    # the contig we want to search on
    })

```

**Comment by jeromekelleher** [comment]
Nice catch. Can you have a look at this please @dcolligan?

**Comment by dcolligan** [comment]
I get a 200 with an empty response.  If the referenceName does not exist, it fails the `referenceName in self._chromFileMap` test in `getVariants` and falls through to the end of the method.

What should be the behavior here?  Should we throw an error if the referenceName is not found in the variantSet?  Should we attempt to convert to a string if we receive an integer referenceName argument?  Should we throw an error if referenceName is not a string?

**Comment by jeromekelleher** [comment]
Is server input validation turned on here? Typing should be completely unambiguous; if the input JSON isn't the right type, we return an error, simple as. We should turn it on if it's still disabled by default, as the only reason it was turned off was because the old compliance tests sent malformed JSON (sigh).

It's not clear what the correct semantics for failing a 'name' search should be. I don't think there's much point in getting worried about it though, as it should really be ``referenceId``, and the schemas should be changed. Then I think it's clear (return 404).

**Comment by dcolligan** [comment]
Yes, validation is turned off for development by default.  I'll change that in the PR.

**Comment by dcolligan** [comment]
Fixed in #943

**Comment by jeromekelleher** [comment]
Out of curiosity, what is the HTTP error that is sent back now?

**Comment by diekhans** [comment]
We want the behavior to be exactly what is documented in the
schema ;-)

The check would be for the referenceName being in the
referenceSet.  Unclear if we actually want to do this.  It's
legitimate for a variantset to have no data on a reference.

The problem is that referenceName should be an string, not and
integer.  It's "1", 1 is violation of the schema.

The decoding of json is going to completely change with protocol
buffers, so I would suggest waiting until protobug is integrated
to address this bug.

**Comment by diekhans** [comment]
500

**Comment by jeromekelleher** [comment]
OK, 500 is wrong, it should be 400 (BAD_REQUEST).

**Comment by david4096** [comment]
This is the error with validation turned on

```
127.0.0.1 - - [24/Feb/2016 15:09:00] "POST /variants/search HTTP/1.1" 200 -
{u'start': 1, u'variantSetId': u'YnJjYTE6MWtnUGhhc2Uz', u'end': 100000000, u'referenceName': 1}
--------------------------------------------------------------------------------
ERROR in app [/usr/local/lib/python2.7/site-packages/flask/app.py:1423]:
Exception on /variants/search [POST]
--------------------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/flask/app.py", line 1475, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python2.7/site-packages/flask/app.py", line 1461, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/Users/david/server-1/ga4gh/frontend.py", line 502, in searchVariants
    flask.request, app.backend.runSearchVariants)
  File "/Users/david/server-1/ga4gh/frontend.py", line 402, in handleFlaskPostRequest
    return handleHttpPost(flaskRequest, endpoint)
  File "/Users/david/server-1/ga4gh/frontend.py", line 278, in handleHttpPost
    responseStr = endpoint(request.get_data())
  File "/Users/david/server-1/ga4gh/backend.py", line 664, in runSearchVariants
    self.variantsGenerator)
  File "/Users/david/server-1/ga4gh/backend.py", line 484, in runSearchRequest
    self.validateRequest(requestDict, requestClass)
  File "/Users/david/server-1/ga4gh/backend.py", line 261, in validateRequest
    jsonDict, requestClass)
  File "/Users/david/server-1/ga4gh/exceptions.py", line 151, in __init__
    jsonDict, requestClass, validator.getInvalidFields(jsonDict))
  File "/Users/david/server-1/ga4gh/avrotools.py", line 57, in getInvalidFields
    invalidFields = validator.getInvalidFields(jsonDict)
  File "/Users/david/server-1/ga4gh/avrotools.py", line 213, in getInvalidFields
    fields = self.handleSchema(self.schema, jsonDict)
  File "/Users/david/server-1/ga4gh/avrotools.py", line 172, in handleSchema
    return self.handleSchemaDispatch(schema, handler, extra)
  File "/Users/david/server-1/ga4gh/avrotools.py", line 207, in handleSchemaDispatch
    value = handler(schema, extra)
  File "/Users/david/server-1/ga4gh/avrotools.py", line 336, in handleRecord
    raise AvrotoolsException(key)
AvrotoolsException: callSetIds
```

**Comment by jeromekelleher** [comment]
Thanks @david4096; over to you @dcolligan!

**Comment by david4096** [comment]
It looks like turning on validation is causing 500s for other queries as well.

If I don't specify a variant set ID I get this:

```
--------------------------------------------------------------------------------
ERROR in app [/usr/local/lib/python2.7/site-packages/flask/app.py:1423]:
Exception on /variants/search [POST]
--------------------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/flask/app.py", line 1475, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python2.7/site-packages/flask/app.py", line 1461, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/Users/david/server-1/ga4gh/frontend.py", line 508, in searchVariants
    flask.request, app.backend.runSearchVariants)
  File "/Users/david/server-1/ga4gh/frontend.py", line 408, in handleFlaskPostRequest
    return handleHttpPost(flaskRequest, endpoint)
  File "/Users/david/server-1/ga4gh/frontend.py", line 285, in handleHttpPost
    responseStr = endpoint(request.get_data())
  File "/Users/david/server-1/ga4gh/backend.py", line 818, in runSearchVariants
    self.variantsGenerator)
  File "/Users/david/server-1/ga4gh/backend.py", line 620, in runSearchRequest
    self.validateRequest(requestDict, requestClass)
  File "/Users/david/server-1/ga4gh/backend.py", line 366, in validateRequest
    jsonDict, requestClass)
  File "/Users/david/server-1/ga4gh/exceptions.py", line 151, in __init__
    jsonDict, requestClass, validator.getInvalidFields(jsonDict))
  File "/Users/david/server-1/ga4gh/avrotools.py", line 57, in getInvalidFields
    invalidFields = validator.getInvalidFields(jsonDict)
  File "/Users/david/server-1/ga4gh/avrotools.py", line 213, in getInvalidFields
    fields = self.handleSchema(self.schema, jsonDict)
  File "/Users/david/server-1/ga4gh/avrotools.py", line 172, in handleSchema
    return self.handleSchemaDispatch(schema, handler, extra)
  File "/Users/david/server-1/ga4gh/avrotools.py", line 207, in handleSchemaDispatch
    value = handler(schema, extra)
  File "/Users/david/server-1/ga4gh/avrotools.py", line 336, in handleRecord
    raise AvrotoolsException(key)
AvrotoolsException: variantSetId
```

**Comment by dcolligan** [comment]
RequestValidationFailureException is currently set to return a 400.  What looks like it's happening in these stacktraces is that in the process of trying to send back an intelligent error, avrotools chokes and the error becomes a 500.

**Comment by david4096** [comment]
So is the bug that request validation is not correct? It's a spurious error for me because I can send in requests with strange keys:

For example, not the `awesome`:

    {u'start': 41196312, u'variantSetId': u'YnJjYTE6YnJjYTE', u'end': 41198948, u'awesome': u'cows', u'referenceName': u'17'}

This request is handled without any complaint.

Then here is a a malformed request where the user accidentally encodes an array (`callSetIds`) as a string.

    {u'start': 41196312, u'variantSetId': u'YnJjYTE6YnJjYTE', u'callSetIds': u'["YnJjYTE6YnJjYTE6SEcwMjEyOA"]', u'end': 41199000, u'referenceName': u'17'}

I get a 500 with this dump:

```
--------------------------------------------------------------------------------
ERROR in app [/usr/local/lib/python2.7/site-packages/flask/app.py:1423]:
Exception on /variants/search [POST]
--------------------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/flask/app.py", line 1475, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python2.7/site-packages/flask/app.py", line 1461, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/Users/david/server-1/ga4gh/frontend.py", line 502, in searchVariants
    flask.request, app.backend.runSearchVariants)
  File "/Users/david/server-1/ga4gh/frontend.py", line 402, in handleFlaskPostRequest
    return handleHttpPost(flaskRequest, endpoint)
  File "/Users/david/server-1/ga4gh/frontend.py", line 278, in handleHttpPost
    responseStr = endpoint(request.get_data())
  File "/Users/david/server-1/ga4gh/backend.py", line 664, in runSearchVariants
    self.variantsGenerator)
  File "/Users/david/server-1/ga4gh/backend.py", line 484, in runSearchRequest
    self.validateRequest(requestDict, requestClass)
  File "/Users/david/server-1/ga4gh/backend.py", line 261, in validateRequest
    jsonDict, requestClass)
  File "/Users/david/server-1/ga4gh/exceptions.py", line 151, in __init__
    jsonDict, requestClass, validator.getInvalidFields(jsonDict))
  File "/Users/david/server-1/ga4gh/avrotools.py", line 57, in getInvalidFields
    invalidFields = validator.getInvalidFields(jsonDict)
  File "/Users/david/server-1/ga4gh/avrotools.py", line 213, in getInvalidFields
    fields = self.handleSchema(self.schema, jsonDict)
  File "/Users/david/server-1/ga4gh/avrotools.py", line 172, in handleSchema
    return self.handleSchemaDispatch(schema, handler, extra)
  File "/Users/david/server-1/ga4gh/avrotools.py", line 207, in handleSchemaDispatch
    value = handler(schema, extra)
  File "/Users/david/server-1/ga4gh/avrotools.py", line 336, in handleRecord
    raise AvrotoolsException(key)
AvrotoolsException: pageSize
```

I think the original bug is not handled and we've cropped up some problems with `avrotools.py`.

**Comment by dcolligan** [comment]
If validation fails, avrotools is called to provide a more useful error (see `exceptions.py`).  Avrotools is choking, and producing the 500.

The reason for not getting an error when adding extra fields to the json dict is (as I just found out) avro's `validate` method does not care about extra fields being present in a record.  Which... seems wrong.

**Comment by dcolligan** [comment]
With regards to extra fields, see #951

**Comment by jeromekelleher** [comment]
#952 is merged; can you check if this fixes this specific bug @david4096, and close if so? We don't want to get too tied up on putting more band-aids over Avro's gaping flaws...

**Comment by dcolligan** [comment]
Fixed in #952

**Comment by david4096** [comment]
Now my request 400s instead of 500s, but it shouldn't.

**Comment by jeromekelleher** [comment]
Isn't 400 the correct response? Client sends garbage, we say 'no thanks'?

**Comment by david4096** [comment]
It doesn't respect optional fields I think.
```
RequestValidationFailureException: Request '{u'start': u':0', u'variantSetId': u'YnJjYTE6YnJjYTE', u'end': 400, u'referenceName': u'1'}' is not a valid instance of <class 'ga4gh._protocol_definitions.SearchVariantsRequest'>; invalid fields: {u'start': u':0', u'callSetIds': u'MISSING_VALUE', u'pageSize': u'MISSING_VALUE', u'pageToken': u'MISSING_VALUE'}
```

**Comment by jeromekelleher** [comment]
OK, can we fix this up please @dcolligan? Let's just go back to ignoring stuff we don't recognise and correctly accepting unset optional fields. We don't want to get bogged down on this, it's really not that important what we do in these error cases, as all of this will be dumped pretty soon.

**Comment by dcolligan** [comment]
Works for me:
```
$ curl -X POST -H "Content-Type: application/json" -d '{"start": 0, "variantSetId": "MWtnLXAzLXN1YnNldDptdm5jYWxs", "end": 400, "referenceName": "1"}' http://localhost:8000/variants/search
{"nextPageToken": null,"variants": []}
```

**Comment by dcolligan** [comment]
Er, I see.  You're saying when it fails for reason X additionally it reports optional fields.  If the request succeeds otherwise (no optional fields specified), things work just fine.

**Comment by jeromekelleher** [comment]
OK, now I'm confused... If this works correctly when we have correct input (including cases when optional fields are not provided), and sort-of works when we have bad input, I think we should move on.

**Comment by dcolligan** [comment]
Fixed in #956
